src/bin/miri.rs

```diff
@@ -63,6 +85,34 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
     }
 }
 
+struct OverrideExportedSymbols {
+    target_crate: bool,
+}
+
+impl rustc_driver::Callbacks for OverrideExportedSymbols {
+    fn config(&mut self, config: &mut Config) {
+        if config.opts.prints.is_empty() && self.target_crate {
+            config.override_queries = Some(|_, local_providers, _| {
```

Comment:
I'm a bit confused, are there two ways to override queries? Why is one query overriden in `MiriCompilerCalls` and another one here?

Reply:
This is to override it for dependencies, i.e. when `MIRI_BE_RUSTC=target` is set. The one in `MiriCompilerCalls` is to override when `MIRI_BE_RUSTC` is not set.

Reply:
Oh I see. Maybe call it `MiriBeRustCompilerCalls` then or so?

Why do we have to even change the behavior of those rustc runs...? The intention was that this would literally be exactly like rustc. (We used to actually invoke `rustc` instead, but finding the right rustc binary turned out to be tricky.)

Reply:
> Why do we have to even change the behavior of those rustc runs...?

Because it originally does not dump the information (`exported_symbols`) Miri want in the crate metadata in `!tcx.sess.opts.output_types.should_codegen()` mode, at least I didn't find a way to retrieve the information.

Reply:
Ah so this affects the information stored in the generated rlibs? Interesting. Please explain that in a comment.